#! /bin/sh /usr/share/dpatch/dpatch-run
## 01_dbus_gconf.dpatch by Javier Uruen Val <javi@warp.es>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch to support the new gconf version using dbus

@DPATCH@

diff -Naur libebox-hardy/src/EBox.pm libebox-intrepid/src/EBox.pm
--- libebox-hardy/src/EBox.pm	2008-05-12 19:10:26.000000000 +0200
+++ libebox-intrepid/src/EBox.pm	2008-08-23 14:44:23.000000000 +0200
@@ -20,8 +20,10 @@
 
 use EBox::Config;
 use EBox::Exceptions::DeprecatedMethod;
-use POSIX qw(setuid setgid setlocale LC_ALL LC_NUMERIC);
 use English;
+use POSIX qw(setuid setgid setlocale LC_ALL LC_NUMERIC);
+
+use constant DBUS_CMD => 'ebox-dbus-launch';
 
 my $loginit = 0;
 
@@ -119,6 +121,40 @@
 	my $user = EBox::Config::user();
 	my $uid = getpwnam($user);
 	setuid($uid) or die "Cannot change user to $user";
+        dbusInit();
+}
+
+# Method: dbusInit
+#
+#      Initialise a dbus daemon, if it's not already done. We store
+#      one for root and one for eBox user.
+#
+#
+sub dbusInit
+{
+    my $confFile;
+    if ( POSIX::getuid() == 0) {
+        $confFile = EBox::Config::conf() . 'dbus-root-session.conf';
+    } else {
+        $confFile = EBox::Config::conf() . 'dbus-ebox-session.conf';
+    }
+    my ($dbusAddress, $dbusDaemonPid, $launchNew) = (0, 0, 1);
+
+    if ( -r $confFile ) {
+        $dbusAddress = EBox::Config::configkeyFromFile(
+			'DBUS_SESSION_BUS_ADDRESS', $confFile);
+        $dbusDaemonPid = EBox::Config::configkeyFromFile(
+			'DBUS_SESSION_BUS_PID', $confFile);
+    }
+
+    # TODO: dbus-send would be cooler than kill
+    unless ( $dbusDaemonPid and (kill 0, $dbusDaemonPid) ) {
+        system( EBox::Config::pkgdata() .  DBUS_CMD . " $confFile");
+        chmod(0660, $confFile);
+        $dbusAddress = EBox::Config::configkeyFromFile('DBUS_SESSION_BUS_ADDRESS', $confFile);
+    }
+
+    $ENV{DBUS_SESSION_BUS_ADDRESS} = $dbusAddress;
 }
 
 1;
